python/plugins/processing/algs/qgis/CreateAtlasGrid.py

```diff
@@ -0,0 +1,178 @@
+        self.addParameter(QgsProcessingParameterDistance(self.COVERAGE_BUFFER,
+                                                         self.tr('Coverage Buffer'), parentParameterName='COVERAGE_LAYER',
+                                                         defaultValue=0.0))
+        self.addParameter(QgsProcessingParameterBoolean(self.ONLY_ON_FEATURES,
+                                                        self.tr('Only on intersection features'),
+                                                        defaultValue=True))
+        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT,
+                                                            self.tr('Grid'),
+                                                            QgsProcessing.TypeVectorPolygon))
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        layout_name = self.parameterAsString(
+            parameters,
+            self.LAYOUT,
+            context
+        )
+        coverage_layer = self.parameterAsVectorLayer(
+            parameters,
+            self.COVERAGE_LAYER,
+            context
+        )
+        coverage_distance = self.parameterAsDouble(
+            parameters,
+            self.COVERAGE_BUFFER,
+            context
+        )
+        only_on_features = self.parameterAsBool(
+            parameters,
+            self.ONLY_ON_FEATURES,
+            context
+        )
+
+        fields = QgsFields()
+        fields.append(QgsField('id', QVariant.Int, '', 10, 0))
+        (sink, dest_id) = self.parameterAsSink(
+            parameters,
+            self.OUTPUT,
+            context,
+            fields,
+            QgsWkbTypes.Polygon,
+            coverage_layer.sourceCrs()
+        )
+
+        if sink is None:
+            raise QgsProcessingException(self.invalidSinkError(parameters, self.OUTPUT))
+
+        layout = QgsProject().instance().layoutManager().layoutByName(layout_name)
+        map = layout.referenceMap()
+        h = map.extent().height()
+        w = map.extent().width()
+
+        coverage = coverage_layer.sourceExtent().buffered(coverage_distance)
+
+        xmin = coverage.xMinimum()
+        xmax = coverage.xMaximum()
+        resto = (w - ((xmax - xmin) % w)) / 2
+
+        yMax = coverage.yMaximum()
+        yMin = yMax - h
+
+        n = 1
+        current_y = 0
+        # Columns loop
+        while current_y < coverage.height():
+            current_x = 0
+            xMin = coverage.xMinimum() - resto
+            xMax = xMin + w
+            # Lines loop
+            while current_x < coverage.width():
+                geom = QgsRectangle(xMin, yMin, xMax, yMax)
+                output_feature = QgsFeature()
+                output_feature.setGeometry(QgsGeometry.fromRect(geom))
+                output_feature.setAttributes([n])
+                if only_on_features:
+                    if len(list(coverage_layer.getFeatures(geom))) > 0:
```

Comment:
I would set getFeatures to avoid to get attributes and geometry to avoid to fetch these data only to check an intersection. e.g set a getFeature with a QgsFeatureRequest set to avoid get data.

Reply:
Good idea

Reply:
You'll also need to use the exact intersection flag in your feature request, or only the bounding boxes are considered. I'd also set a limit of (1) in the feature request as you only care if 0 or >= 1 features are found.